src/thorin/be/llvm/amdgpu_pal.cpp

```diff
@@ -0,0 +1,87 @@
+#include "thorin/be/llvm/amdgpu_pal.h"
+
+#include <unordered_map> // TODO don't use std::unordered_*
+
+#include "thorin/primop.h"
+#include "thorin/world.h"
+
+namespace thorin::llvm {
+
+AMDGPUPALCodeGen::AMDGPUPALCodeGen(World& world, const Cont2Config& kernel_config, int opt, bool debug)
+    : CodeGen(world, llvm::CallingConv::AMDGPU_Gfx, llvm::CallingConv::C, llvm::CallingConv::AMDGPU_CS, opt, debug)
+    , kernel_config_(kernel_config)
+{
+    module().setDataLayout("e-p:64:64-p1:64:64-p2:32:32-p3:32:32-p4:64:64-p5:32:32-p6:32:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64-S32-A5-G1-ni:7");
+    module().setTargetTriple("amdgcn-amd-amdpal");
+}
+
+//------------------------------------------------------------------------------
+// Kernel code
+//------------------------------------------------------------------------------
+
+void AMDGPUPALCodeGen::emit_fun_decl_hook(Continuation* continuation, llvm::Function* f) {
```

Comment:
The following functions:
 * `emit_fun_decl_hook`
 * `emit_global`
 * `emit_mathop`
 * `emit_reserve`
 
appear identical between this and the HSA path. I strongly recommend keeping `amdgpu.[h/cpp]` as abstract classes and moving the shared code there.